processing/src/main/java/org/apache/druid/query/aggregation/first/NumericFirstVectorAggregator.java

```diff
@@ -0,0 +1,174 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.first;
+
+import org.apache.druid.common.config.NullHandling;
+import org.apache.druid.query.aggregation.VectorAggregator;
+import org.apache.druid.segment.vector.VectorValueSelector;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+/**
+ * Class for vectorized version of first/earliest aggregator over numeric types
+ */
+public abstract class NumericFirstVectorAggregator implements VectorAggregator
+{
+  static final int NULL_OFFSET = Long.BYTES;
+  static final int VALUE_OFFSET = NULL_OFFSET + Byte.BYTES;
+  final VectorValueSelector valueSelector;
+  private final boolean useDefault = NullHandling.replaceWithDefault();
+  private final VectorValueSelector timeSelector;
+  private long firstTime;
+
+  public NumericFirstVectorAggregator(VectorValueSelector timeSelector, VectorValueSelector valueSelector)
+  {
+    this.timeSelector = timeSelector;
+    this.valueSelector = valueSelector;
+    firstTime = Long.MAX_VALUE;
+  }
+
+  @Override
+  public void init(ByteBuffer buf, int position)
+  {
+    buf.putLong(position, Long.MAX_VALUE);
+    buf.put(position + NULL_OFFSET, useDefault ? NullHandling.IS_NOT_NULL_BYTE : NullHandling.IS_NULL_BYTE);
+    initValue(buf, position + VALUE_OFFSET);
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position, int startRow, int endRow)
+  {
+    final long[] timeVector = timeSelector.getLongVector();
+    final boolean[] nullTimeVector = timeSelector.getNullVector();
+    final boolean[] nullValueVector = valueSelector.getNullVector();
+    firstTime = buf.getLong(position);
+
+    // the time vector is already sorted
+    // if earliest is on the default time dimension
+    // but if earliest uses earliest_by it might use a secondary timestamp
+    // which is not sorted. For correctness, we need to go over all elements.
+    // A possible optimization here is to have 2 paths one for earliest where
+    // we can take advantage of the sorted nature of time
+    // and the earliest_by where we have to go over all elements.
+    int index;
+
+    for (int i = startRow; i < endRow; i++) {
+      index = i;
+      if (nullTimeVector != null && nullTimeVector[index]) {
+        continue;
```

Comment:
how does this work if all of the time values are null? The docs seem to indicate that if the time column is null we just take the first value. I wonder if we should we maybe treat rows with a null timestamp as if the timestamp were Long.MAX_VALUE and update the value? Though, I guess in that case we would have trouble distinguishing null time and null row from the initialized state. I suppose we could be consistent if we made a more general behavior that if there are multiple values with the same timestamp we take the first non-null value we encounter, though I suspect that would require change with the non-vector aggs too.

The more I think about stuff, I think it was a mistake that first/last allow null values to be first/last... but I don't think it is appropriate to change that in this PR.

Reply:
So currently in case of non-null timestamps, we just take the first value. We iterate over the values and only update the first time if the current time is less than the earliest time. The issue arises when all timestamps are null. Considering the earliest where the time selector is __time, the chances of this happening are more when used on a secondary timestamp through earliest_by. In such a case, should we even return any results ? The docs point that ```
 If expr comes from a relation with a timestamp column (like __time in a Druid datasource), the "earliest" is taken from the row with the overall earliest non-null value of the timestamp column.
 ```